[ddp] parameter verification

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74127
* __->__ #74113

Check mismatch in # of parameters by broadcasting and verifying from rank 0. As a result, non-zero ranks raise an error when # of parameters are mismatched across ranks.

Closes https://github.com/pytorch/pytorch/issues/73547

Differential Revision: [D34772067](https://our.internmc.facebook.com/intern/diff/D34772067/)

rohan-varma committed:
[ddp] parameter verification

Check mismatch in # of parameters by broadcasting and verifying from rank 0. As a result, non-zero ranks raise an error when # of parameters are mismatched across ranks.

Closes https://github.com/pytorch/pytorch/issues/73547

Differential Revision: [D34772067](https://our.internmc.facebook.com/intern/diff/D34772067/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[ddp] parameter verification"


Check mismatch in # of parameters by broadcasting and verifying from rank 0. As a result, non-zero ranks raise an error when # of parameters are mismatched across ranks.

Closes https://github.com/pytorch/pytorch/issues/73547

Differential Revision: [D34772067](https://our.internmc.facebook.com/intern/diff/D34772067/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[ddp] parameter verification"


Check mismatch in # of parameters by broadcasting and verifying from rank 0. As a result, non-zero ranks raise an error when # of parameters are mismatched across ranks.

Closes https://github.com/pytorch/pytorch/issues/73547

Differential Revision: [D34772067](https://our.internmc.facebook.com/intern/diff/D34772067/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[ddp] parameter verification"


Check mismatch in # of parameters by broadcasting and verifying from rank 0. As a result, non-zero ranks raise an error when # of parameters are mismatched across ranks.

Closes https://github.com/pytorch/pytorch/issues/73547

Differential Revision: [D34772067](https://our.internmc.facebook.com/intern/diff/D34772067/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -2052,6 +2052,34 @@ void verify_params_across_processes(
+
+  // Broadcast and verify parameter size.
+  std::vector<at::Tensor> param_size_vec{param_size_tensor};
+  process_group->broadcast(param_size_vec)->wait();
```
nit (can be done as future BE work item): DDP ctor will broadcast param values anyway. Is it possible to implement these two broadcasts in the same location using two async ops? I recall the param value broadcast was done in the `distributed.py` file?

rohan-varma replied:
Yes, there's actually 3 collectives that go on now:

1. parameter size verification
2. parameter shape verification
3. parameter broadcast

We could make these all async, and wait on all three, and register then() callbacks on the first two to raise appropriate errors. Filed https://github.com/pytorch/pytorch/issues/74185

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -2052,6 +2052,34 @@ void verify_params_across_processes(
+  std::vector<at::Tensor> param_size_vec{param_size_tensor};
+  process_group->broadcast(param_size_vec)->wait();
+  auto res = param_size_tensor[0].item<int>();
+  TORCH_CHECK(
```
curious, does this mean some DDP process will crash, while other won't if their size matches? If that's the case, does it mean allgather might be better, as all processes can make the same decision here and throw the same error?

rohan-varma replied:
This is a good point, it would be a lot better if all processes make the same decision. Next update will fix this

mrshenli commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -7228,39 +7237,73 @@ def test_verify_model_across_rank_with_logger(self):
+            group_gloo = dist.new_group(
+                timeout=timedelta(seconds=60), backend=dist.Backend.GLOO
+            )
+            # Set NCCL_BLOCKING_WAIT and use a new NCCL group to improve test
```
curious, why this can help improve determinism?

rohan-varma replied:
We'd like NCCL_BLOCKING_WAIT so that rank 0 doesn't need to be taken down by async error handling which is non-deterministic when it kicks in and runs, and it also takes down the process which is hard to accomodate for in unittest.

As a result, we also need a new NCCL group as the default NCCL group is initialized before this test executes as a part of setup. I'll add some clarifying comments.

mrshenli reviewed (approved):
Stamp to unblock

rohan-varma committed:
Update on "[ddp] parameter verification"


Check mismatch in # of parameters by broadcasting and verifying from rank 0. As a result, non-zero ranks raise an error when # of parameters are mismatched across ranks.

Closes https://github.com/pytorch/pytorch/issues/73547

Differential Revision: [D34772067](https://our.internmc.facebook.com/intern/diff/D34772067/)

[ghstack-poisoned]
